src/authoring/hooks/use-curriculum.ts

```diff
@@ -11,31 +13,41 @@ export const units = ["cas", "mods", "brain", "m2s"];
 export const branches = ["authoring-testing"];
 
 export const useCurriculum = (auth: Auth, api: AuthoringApi) => {
-  const [authState, setAuthState] = useState<AuthState>("authenticated");
-  const [branch, _setBranch] = useState<string | undefined>(undefined);
-  const [unit, _setUnit] = useState<string | undefined>(undefined);
-  const [path, setPath] = useState<string | undefined>(undefined);
-  const [unitConfig, setUnitConfig] = useState<IUnit | undefined>(undefined);
-  const [files, setFiles] = useState<IUnitFiles | undefined>(undefined);
-  const [error, setError] = useState<string | undefined>(undefined);
+  const [authState, setAuthState] = useImmer<AuthState>("authenticated");
+  const [branch, _setBranch] = useImmer<string | undefined>(undefined);
+  const [unit, _setUnit] = useImmer<string | undefined>(undefined);
+  const [path, setPath] = useImmer<string | undefined>(undefined);
+  const [unitConfig, _setUnitConfig] = useImmer<IUnit | undefined>(undefined);
+  const [files, setFiles] = useImmer<IUnitFiles | undefined>(undefined);
+  const [error, setError] = useImmer<string | undefined>(undefined);
   const lastUnitRef = useRef<string | undefined>(undefined);
   const filesRef = useRef<firebase.database.Reference | undefined>(undefined);
+  const [saveState, setSaveState] = useImmer<SaveState | undefined>(undefined);
+  const saveUnitConfigRef = useRef(false);
+  const saveStateClearTimeoutRef = useRef<number>();
 
-  const reset = () => {
+  const reset = useCallback(() => {
     setError(undefined);
     _setBranch(undefined);
     _setUnit(undefined);
-    setUnitConfig(undefined);
+    _setUnitConfig(undefined);
+  }, [_setBranch, _setUnit, setError, _setUnitConfig]);
+
+  // externally when setUnitConfig is called, we want to update the state
+  // and also call the api to save the changes
+  const setUnitConfig: Updater<IUnit | undefined> = (draft) => {
+    saveUnitConfigRef.current = true;
```

Comment:
I'm not sure if it makes sense yet to switch. But this idea of responding to changes to the unit and then saving it, is what we use MST for in CLUE and CODAP. And we already should have MST models defined for the unit configuration (I think). If we have those models already then it seems useful to use those models and MST for monitoring changes to the config.

That approach would also separate the saving of the config from the re-rendering of the component that is calling useCurriculum. Currently its necessary for the component to re-render each time the config is changed so that it can be saved.